sage.py

```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+    print('Victims hosts: ', set([team_victim.split('-')[-1] for team_victim in victim_episodes.keys()]))
+
+    attacker_episodes = dict()  # Episodes per (team, attacker)
+    for attack, episodes in state_sequences.items():
+        team = attack.split('-')[0]
+        attacker = (attack.split('->')[0]).split('-')[1]
+        team_attacker = team + '-' + attacker
+        if team_attacker not in attacker_episodes.keys():
+            attacker_episodes[team_attacker] = []
+        attacker_episodes[team_attacker].extend(episodes)
+        attacker_episodes[team_attacker] = sorted(attacker_episodes[team_attacker], key=lambda epi: epi[0])
+    print('Attacker hosts: ', set([team_attacker.split('-')[1] for team_attacker in attacker_episodes.keys()]))
+
+    return attacker_episodes, victim_episodes
+
+
+# Translate technical nodes to human-readable
+def _translate(label, root=False):
     new_label = ""
     parts = label.split("|")
     if root:
-        new_label += 'Victim: '+str(root)+'\n'
+        new_label += 'Victim: ' + str(root) + '\n'
 
     if len(parts) >= 1:
         new_label += verbose_micro[parts[0]]
     if len(parts) >= 2:
-        new_label += "\n"+parts[1]
+        new_label += "\n" + parts[1]
     if len(parts) >= 3:
-        new_label += " | ID: "+parts[2]
+        new_label += " | ID: " + parts[2]
 
     return new_label
-    
-## Per-objective attack graph for dot: 14 Nov (final attack graph) 
-def make_AG(condensed_v_data, condensed_data, state_groups, sev_sinks, datafile, expname):  
-    tcols = {
-        't0': 'maroon',
-        't1': 'orange',
-        't2': 'darkgreen',
-        't3': 'blue',
-        't4': 'magenta',
-        't5': 'purple',
-        't6': 'brown',
-        't7': 'tomato',
-        't8': 'turquoise',
-        't9': 'skyblue',
-        
-    }
+
+
+def _get_objective_nodes(state_sequences, obj_only=False):
+    objectives = set()
+    for episodes in state_sequences.values():  # Iterate over all episodes and collect the objective nodes
+        for epi in episodes:
+            if len(str(epi[2])) == 3:  # If high-severity, then include it
+                mcat = micro[epi[2]].split('.')[1]
+                if obj_only:                         # Experiment: only mcat or mcat + mserv?
+                    vert_name = mcat
+                else:
+                    vert_name = mcat + '|' + epi[4]
+                objectives.add(vert_name)
+    return list(objectives)
+
+
+def _is_severe_sink(vname, sev_sinks):
+    for sink in sev_sinks:
+        if vname.split("|")[-1] == sink:
+            return True
+    return False
+
+
+def _make_vertex_info(episode, in_main_model):
+    start_time = round(episode[0] / 1.0)
+    end_time = round(episode[1] / 1.0)
+    cat = micro[episode[2]].split('.')[1]
+    signs = episode[5]
+    timestamps = episode[6]
+    if episode[3] in in_main_model:  # TODO: this check is useless, since it is always True
```

Comment:
Here is another discussion. `in_main_model` seems useless, since it is created from the state IDs from the episodes and then the same states are queried in `in_main_model` (so all states will be in `in_main_model`, including the ones with ID `-1`). So, all the if-checks for `|Sinks` are never true and are currently not used. What could be checked instead is `if episode[3] != -1`, since nodes with ID -1 are low-severity sinks (I think, but also see the comment above in the traverse method). This will lead to some states having `|Sink` ID (which I saw on the image in some of the papers, but which is not the case for the current implementation), and as a result a bit more states (these states will now have an ID "Sink" and hence will not be merged with a non-sink nodes without ID (e.g. some VulnD).

I am not sure which solution you would prefer, but this has to be handled in a separate PR (since it changes the resulting AGs). If something from what I have written is unclear, we can also discuss it later

Reply:
Lets discuss this. Ideally, I want all sinks states to have |Sink, even the high-sev ones that we salvaged. I think this is the only solution to the context problem (where multiple contextually different objectives are shown but some are directly from sinks so no learning was applied).

Reply:
Hmm, not sure about this. If there are two mid-sev sinks and they have ID `Sink`, then graphviz will merge them as identical nodes, *no matter where they appear in the attack path and what is their context*. So, yes, this can be discussed

Reply:
Created #27 for it

Reply:
Thank you! I just wanted to make sure: is it an issue "for later"? In the sense that, do I still replace `if episode[3] in_main_model` with `if episode[3] != -1` (in a separate issue, since it will change the AGs)?

Reply:
The code can stay like it is for now. The issue we created can deal with it later.